Add support for providing custom registries

stefanmendoza opened the pull request:
### What was wrong?

https://github.com/ethereum/web3.py/issues/829

Specifically [this comment](https://github.com/ethereum/web3.py/issues/829#issuecomment-438387257) and [this comment](https://github.com/ethereum/web3.py/issues/829#issuecomment-438424600). This in the first step towards supporting tuple types in web3.py - allow for providing customer registries to allow for customer encoders and decoders.

### How was it fixed?

Added support for providing custom registries to the `eth_abi.abi` methods. The methods assume the `eth_abi.registry.registry` is the default.

#### Cute Animal Picture

![Cute animal picture](https://img-s-msn-com.akamaized.net/tenant/amp/entityid/BBKHUZJ.img)

davesque commented:
Hey, Stefan.  Thanks for your work on this.  Some of our team members had a previous discussion about fixing this issue.  In that discussion, we actually decided that the best path forward would be to convert `encode_single`, `encode_abi`, etc. into methods on the `eth_abi.registry.ABIRegistry` class.  That way, there's no need to pass around the active registry as an argument and it would always be available as `self` in the context of registry operations.

stefanmendoza commented:
@davesque okay, sounds good! Since that's a non-passive change, is the expectation that we would mark the `eth_abi.abi` methods as deprecated or just remove them since we're still on the beta and add a deprecation warning in the 1.x branch?

davesque commented:
The methods could still exist in the same location.  It's just that, instead of being stand-alone functions, they would be references to bound methods on a class instance.  For example, you could end up doing something like this:
```python
# eth_abi/abi.py

from .registry import registry

encode_single = registry.encode_single
encode_abi = registry.encode_abi
...
```
However, I'm realizing now that this doesn't really solve the issue of how web3.py or any other external consumer of `eth_abi` knows what registry to use.  I'll have to spend some more time thinking about that one.

stefanmendoza commented:
@davesque one thing @carver and I were discussing yesterday was potentially having a Codec class and passing that around as naming wise it makes more sense, you're not overloading the registry's purpose, and you're passing around the instance you're actually using vs. a registry.

See my comment here and his comment below:
https://github.com/ethereum/web3.py/issues/829#issuecomment-438784925

stefanmendoza committed:
Support custom registries with ABIEncoder

carver commented:
Why did you decide to implement just the encoder instead of the full `ABICodec` concept from the other thread? It requires clients to track both an encoder and decoder pair AFAICT. So without more info, I prefer the paired approach as `ABICodec`.

carver commented on eth_abi/abi.py:
```diff
@@ -23,119 +23,124 @@
-    stream = ContextFramesBytesIO(data)
 
-    return decoder(stream)
+encoder = ABIEncoder()
```
Maybe `default_codec = ABICodec()` instead?

stefanmendoza commented:
@davesque sounds good! Will look into that this week.

@carver thanks for reviewing this. I'll look over your comments this week as well. 👍

stefanmendoza committed:
Revert "Support custom registries with ABIEncoder"

This reverts commit 90aba57138c40cc2672eb547b265be0d299a4095.

stefanmendoza committed:
Create codec module and export methods in eth_abi.abi

stefanmendoza committed:
Created packed codec in eth_abi/codec_packed

davesque commented on eth_abi/codec.py:
```diff
@@ -0,0 +1,146 @@
+
+
+class ABICodec():
+
```
We don't usually leave a blank line like just after a class declaration statement.

davesque commented on eth_abi/codec.py:
```diff
@@ -0,0 +1,146 @@
+
+class ABICodec():
+
+    def __init__(self, registry: ABIRegistry = None):
```
Also, another nitpick.  We typically don't put spaces around the `=` sign for default argument values.

stefanmendoza committed:
Addressing comments

stefanmendoza committed:
address nit about method defaults

carver commented on eth_abi/codec_packed.py:
```diff
@@ -0,0 +1,41 @@
+)
+
+
+class ABICodecPacked(ABICodec):
```
Looks like the inheritance actually works the other way, with some methods only implemented on the non-packed codec.

stefanmendoza replied:
Yeah, it's weird. I feel like the assumption is that any subclass of the `ABIEncoder` would be expected to implement all five methods, but the packed implementation is experimental and isn't decodable. Not sure if it makes sense to have a completely different class for that one. Was trying to keep in line with the ABICodec. Thoughts? I could see both ways.

carver replied:
Another alternative is to make a third base codec class that implements the encoding methods. The packed codec would be like:

```py
class ABICodecPacked(BaseABICodecEncoding):
    pass
```

The full codec would also subclass the Base codec and implement the decoding.

stefanmendoza replied:
@carver I considered that as well, but would we want to make a `BaseABICoderDecoding` as well?

carver replied:
I think it's fine to just implement that in `ABICodec` for now.

stefanmendoza replied:
Updated here: https://github.com/ethereum/eth-abi/pull/109/commits/abceb187b537e3b754356feb3bbfba6a7989684a

stefanmendoza replied:
Added null check for the registry provided to the base encoder: f939865

stefanmendoza replied:
and a test for the null case ^ - d63273c

stefanmendoza replied:
Ended up naming it `BaseABICodecEncoder`. Let me know if you think that doesn't make sense since we have separate `*Encoder` classes. If so, I can make it `BaseABICodecEncoding` like you suggested. Just thought the wording of that was kind of odd. 🤔

carver replied:
Yup, seems fine to me. Could also be `BaseABIEncoder` since it's only responsible for the encoding side of the codec. Whatever you like, since it's not intended to be part of the public API. :)

carver commented on eth_abi/codec_packed.py:
```diff
@@ -0,0 +1,41 @@
+            ABICodec.__init__(self, registry)
+
+    def is_encodable(self, typ: TypeStr, arg: Any) -> bool:
+        raise NotImplementedError("'is_encodable' is not  supported in packed mode")
```
I haven't looked too deeply at the packed stuff. Why can't we test if a value is encodable?

stefanmendoza replied:
Was just keeping with the methods that were previously implemented for packed encoding. I can dig into this some more and see. 🤔

davesque replied:
I think the lack of an `is_encodable_packed` function may have just been an oversight or for the sake of expedience.  I just looked through the code and all of the `Packed*` encoder classes seem to inherit from base classes which implement `validate_value`.  So they should all have what is needed to run an encodability check.  Otherwise, I can't think of another reason that it wouldn't be possible.  Now that we have the `ABICodec` class, we should get this for "free."  We just need delete the `NotImplementedError` override and add something like `is_encodable_packed = default_codec_packed.is_encodable`.

carver replied:
Why rename the method with the `_packed` suffix? We don't rename any of the other methods like `decode_abi_packed`.

I'd say just implement it once in the base codec class and get it for free with the same name.

stefanmendoza replied:
I think he's suggesting that because we're trying to stay passive by exporting the methods in `eth_abi/packed`. Like:
```
encode_abi_packed = default_codec_packed.encode_abi
encode_single_packed = default_codec_packed.encode_single
```
Since those were previously exported.

stefanmendoza replied:
Added the method here:
https://github.com/ethereum/eth-abi/pull/109/commits/abceb187b537e3b754356feb3bbfba6a7989684a#diff-f7c24c7b7a7b6405352611f3ee624887

Need to add some tests now for it.

stefanmendoza committed:
Added support for is_encodable in packed codec

stefanmendoza commented:
Just an update, I think all that's left is adding tests for the Packed ABICodec's is_encodable method and updating the documentation to discuss using custom registries.

stefanmendoza committed:
Added null check on registry in base encoder

stefanmendoza committed:
Added test for null registry to basecodec

stefanmendoza committed:
Added tests for is_encodable with packed registry

stefanmendoza committed:
started adding documentation, fixed sphinx warnings

stefanmendoza commented:
@carver / @davesque - I think this is ready for one last implementation / tests review before I make the final changes (adding the documentation for how to use the codecs with custom registries).

stefanmendoza commented on eth_abi/abi.py:
```diff
@@ -23,119 +23,124 @@
-        encoding e.g. ``'uint256'``, ``'bytes[]'``, ``'(int,int)'``, etc.
-    :param arg: The python value to be encoded.
+    def __init__(self, registry: ABIRegistry = default_registry):
+        self._registry = registry
```
@carver wasn't exactly sure if you meant to make `default_registry` in `eth_abi.registry` into a method? This is how I ended up setting them up

- BaseABICodecEncoder - https://github.com/ethereum/eth-abi/pull/109/commits/abceb187b537e3b754356feb3bbfba6a7989684a#diff-840cea94757e5fecdafdc6c8bb6ba8e2R32
- ABICodec - https://github.com/ethereum/eth-abi/pull/109/commits/abceb187b537e3b754356feb3bbfba6a7989684a#diff-840cea94757e5fecdafdc6c8bb6ba8e2R100
- ABICodecPacked - https://github.com/ethereum/eth-abi/pull/109/commits/abceb187b537e3b754356feb3bbfba6a7989684a#diff-df4a437f066f81aabfd6125511b5e745R18

carver commented on eth_abi/codec.py:
```diff
@@ -100,7 +112,20 @@ def is_encodable(self, typ: TypeStr, arg: Any) -> bool:
+    """
+
     def __init__(self, registry: ABIRegistry=None):
+        """
```
It looks like we can get the `__init__` docs added to the class docs by adding `autoclass_content = 'both'` to the docs conf. Can you experiment with it to see if that works as expected, looks okay, etc?  https://stackoverflow.com/a/9772922/8412986

If so, we will probably add it to our project template so that it eventually gets filters to all our other projects.

carver commented on docs/eth_abi.rst:
```diff
@@ -16,6 +16,22 @@ eth\_abi.base module
+
+.. automodule:: eth_abi.codec
+    :members:
+    :undoc-members:
```
Yeah, I think it's good to have it on. If we don't want something public, it should be `_`-prefixed.

davesque commented:
Yeah, I've got a few commits I'd like to add to this to tie a bow on things.  Will hopefully have that all wrapped up before EOD.

davesque commented:
@stefanmendoza @carver Alright, I made the PR to your fork of `eth-abi`.  Let me know if you have any questions or concerns.

To briefly summarize what I did:
* Made a new article in the docs describing how to use the `ABICodec` and related classes.
* Changed the inheritance scheme and structure of classes in the `codec` module.  Now there is a `BaseABICoder` class which provides the constructor for two new classes: `ABIEncoder` and `ABIDecoder`.  Now, `ABICodec` is just a `pass` statement class which inherits from both of those new classes.

davesque commented:
Here's a link to that PR: https://github.com/stefanmendoza/eth-abi/pull/1

davesque committed:
Web3 issue 829 (#1)

* Move some testing modules

* Add "Codecs" article to docs

Also, remove API docs for `codec_packed` module since it will be
deleted.

* Use common values for encodability tests

* Import lint

* Type annotations and docstring formatting

* Remove explicit checking for `None` registries

* Rearrange inheritance scheme in `codec` module

This will save some code and hopefully add some utility to the classes
provided by that module.

* Add missing __init__.py file

* Typo

* Add note in releases article about codecs

carver commented:
Cool, I'm really happy with this direction! When you write out the docs for a new API, and the examples look straightforward and concise, it's a great confirmation that some good choices were made.
